mate-video-player/data/totem.1
.\" Copyright (C) 2004 Andre Lehovich <andrel@u.arizona.edu>
.\"
.\" This is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2,
.\" or (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License 
.\" along with this; if not write to the Free Software Foundation, Inc.
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
.TH totem 1 "2008\-08\-25" "MATE"
.SH NAME
Totem \- MATE desktop movie player based on GStreamer
.SH SYNOPSIS
.B totem
.RI [ option... ] " " [ filenames | URI ]
.SH DESCRIPTION
.B Totem
is a movie player for the MATE desktop.  Through use of the
.B GStreamer
backend many popular file and streaming formats are supported including:
MPEG audio and video (MP3/MPV), Ogg, AVI, ASF, MOV, WMV,
QuickTime, and more.  Playing of DVD and VCD media is
also possible.
Although primarily intended as a movie player,
.B Totem
can also play many audio formats.
.P
For additional documentation on interactive use of 
.B Totem
select the
.I Help
menu option.  The menus allow access to some features not
available through keybindings or command-line options.
.SH OPTIONS
.TP
.B \-?, --help
Output a longer help message then quit.  (Any discrepancies
between this man page and the output of
.B totem --help
should be reported as a documentation bug.)
.TP
.B --version
Output version information then quit.
.P
The following options command an already-running instance of
.B Totem
to do something; they are useful for remote-control scripting. If 
.B Totem
is not already running, these commands will launch a new instance of
.B Totem
but will not do anything further.
.TP
.B --play-pause
Tell an already-running instance of 
.B Totem
to toggle between play and pause.
.TP
.B --play
Tell an already-running instance of
.B Totem
to play (has no effect if already playing)
.TP
.B --pause
Tell an already-running instance of
.B Totem
to pause (has no effect if already paused)
.TP
.B --next
Tell an already-running instance of 
.B Totem
to skip to the next movie or chapter in the playlist.
.TP
.B --previous
Tell an already-running instance of 
.B Totem
to return to the previous movie or chapter in the playlist.
.TP
.B --seek-fwd
Tell an already-running instance of 
.B Totem
to seek forwards 15 seconds in the current movie.
.TP
.B --seek-bwd
Tell an already-running instance of 
.B Totem
to seek backwards 15 seconds in the current movie.
.TP
.B --volume-up
Tell an already-running instance of 
.B Totem
to raise the volume by 8%.
.TP
.B --volume-down
Tell an already-running instance of 
.B Totem
to lower the volume by 8%.
.TP
.B --fullscreen
Tell an already-running instance of 
.B Totem
to toggle fullscreen mode.
.TP
.B --toggle-controls
Tell an already-running instance of
.B Totem
to toggle showing the controls.
.TP
.B --quit
Tell an already-running instance of
.B Totem
to quit.
.TP
.BI "--enqueue " filename|URI
Tell an already-running instance of Totem to add a new stream
to the playlist.
.TP
.BI "--replace " filename|URI
Tell an already-running instance of Totem to play 
from the playlist.
.SH KEYBINDINGS
Global keybindings for
.BR Totem :
.P
.ta \w'Down-arrow   'u	
a	cycle between aspect ratios
.br
p	toggle between play and pause
.br
Esc	exit full screen mode
.br
f	toggle full screen
.br
h	toggle display of on-screen controls
.br
0	resize window to 50% original size
.br
1	resize window to 100% original size
.br
2	resize window to 200% original size
.br
.br
r	zoom in the video
.br
t	zoom out the video
.br
d	start and stop the telestrator (drawing) mode
.br
e	erase the drawing
.br
Left-arrow        skip back 15 seconds
.br
Right-arrow       skip forward 60 seconds
.br
Shift+Left-arrow  skip back 5 seconds
.br
Shift+Right-arrow skip forward 15 seconds
.br
Ctrl+Left-arrow   skip back 3 minutes
.br
Ctrl+Right arrow  skip forward 10 minutes
.br
Up-arrow	increase volume by 8%
.br
Down-arrow	decrease volume by 8%
.br
b		jump back to previous chapter/movie in playlist
.br
n		jump to next chapter/movie in playlist
.br
q		quit
.br
Ctrl+E          eject the playing optical media
.br
Ctrl+O		open a new file
.br
Ctrl+L		open a new URI
.br
F9		toggle display of the playlist
.br
m		show the DVD menu
.br
c		show the DVD chapter menu
.SH FILES
.I /usr/lib/totem/plugins/
.RS
The global plugins directory, which contains all the system-installed
plugins.
.RE
.I ~/.local/share/totem/plugins/
.RS
The user plugins directory, containing plugins available only to that user.
It is always in the "totem/plugins" subdirectory of your 
.B XDG_DATA_HOME
directory.
.RE
.I ~/.config/totem/state.ini
.RS
The configuration file for the main window's appearance, stored in a
simple INI format.
.RE
.SH BUGS
.B Totem
cannot play files using some proprietary codecs.
.SH AUTHOR
.B Totem
was written by Bastien Nocera <hadess@hadess.net>.
This manual page was written by Andre Lehovich for the
Debian Project.
.SH DISTRIBUTION
The latest version of Totem may be downloaded from
.UR http://projects.mate.org/totem/
<http://projects.mate.org/totem/>
.UE
.SH SEE ALSO
.BR "totem-video-thumbnailer" (1),
.P
The online documentation available through the program's
.I Help
menu.


libmate/NEWS
libmate 2.32.0

mate-sound APIs now use libcanberra internally. That means that
mate_sound_connection_get() and will always return error (-1), and
mate_sound_sample_load also always return -1 but still cache
the sample using the libcanberra API.
libmate continues to link to libesound and libaudiofile for ABI
compatibility; use --disable-esd to remove this dependency.
(#603198, #603349)


libmate 2.30.0

* Enable silent build using AM_SILENT_RULES (Christian Persch)
* Put SlowKeys and BounceKeys default delays back to 300 (Willie Walker)
* Deprecate mate-sound (Christian Persch)
* Deprecate mate_[un]setenv (Christian Persch)
* Update the description of the picture_options key (Matthias Clasen)

* Updated translations
  - Khaled Hosny (ar)
  - astur (ast)
  - Alexander Shopov (bg)
  - Sadia Afroz (bn)
  - Jordi Mallach (ca)
  - Reşat SABIQ (crh)
  - Andre Klapper (cs)
  - Iestyn Pryce (cy)
  - Ask Hjorth Larsen (da)
  - Mario Blättermann (de)
  - Bakaoukas Nikolaos (el)
  - Thomas Thurman (en@shaw)
  - Bruce Cowan (en_GB)
  - Jorge González (es)
  - Mattias Põldaru (et)
  - Iñaki Larrañaga Murgoitio (eu)
  - Tommi Vainikainen (fi)
  - Claude Paroz (fr)
  - Fran Diéguez (gl)
  - Gabor Kelemen (hu)
  - Francesco Marletta (it)
  - Rimas Kudelis (lt)
  - Badral Sanligiin (mn)
  - Kjartan Maraas (nb)
  - Nils-Christoph Fiedler (nds)
  - Wouter Bolsterlee (nl)
  - A S Alam (pa)
  - Piotr Drąg (pl)
  - Duarte Loreto (pt)
  - Henrique P Machado (pt_BR)
  - Adi Roiban (ro)
  - Yuri Kozlov (ru)
  - Marcel Telka (sk)
  - Matej Urbančič (sl)
  - Горан Ракић (sr)
  - Goran Rakić (sr@latin)
  - Daniel Nylander (sv)
  - Maxim Dziumanenko (uk)
  - Clytie Siddall (vi)
  - Chao-Hsiung Liao (zh_HK)
  - Chao-Hsiung Liao (zh_TW)


libmate 2.28.0

Aron Xu (1):
      Corrected 5 strings.
Denis Arnaud (1):
      Updated breton translation and added br in LINGUAS
Lucian Adrian Grijincu (1):
      Updated Romanian translation
Miloš Popović (1):
      Updated Serbian translation
Rodrigo L. M. Flores (1):
      Updated Brazilian Portuguese mailing list address.

 
libmate 2.27.5

* Update default properties for menus_have_icons, buttons_have_icons and
  toolbar_style
* Port mate-sound.c to g_poll to fix portability for old OSX versions without
  poll.
* Update the address of the FSF by replacing old bits with new ones

* Updated translations
  - Jordi Mallach (ca)
  - Jordi Mallach (ca@valencia)
  - Khaled Hosny (ar)
  - Kjartan Maraas (nb)
  - Marcel Telka (sk)
  - Ray Wang (zh_CN)
  - Shankar Prasad (kn)
  - Sweta Kothari (gu)


libmate 2.26.0

* Updated translations
  (as, bn_IN, crh, cs, da, el, gl, hi, it, ja, lt, ml, mr,
   ro, ru, ta, te)

libmate 2.24.1

* Fix warnings about missing format specifiers (Kjartan)
* Remove #if 0'ed out code (Kjartan)
* Updated translations
  (ar, bg, en_GB, eu, fr, gu, lv, nl, or, pl, pt, th, tr,
   vi, zh_HK, zh_TW)

libmate 2.24.0.1

* Make thumbnail housekeeping less agressive by default
  (Michael Chudobiak)

libmate 2.24.0

* Updated translations
  (ar, da, gu, it, ro, ru, sr, ta, te)

libmate 2.23.92

* Port mate-open to GIO. (Jean-Yves Lefort)
* Build fix (Tor Lillqvist)

libmate 2.23.5

* Fixes for sound related issues (Jeff Steadfast)

libmate 2.23.4

* Updated translations.

libmate 2.23.3

* Code cleanups (Kjartan, Christian)
* Added schema to support the thumbnail cache cleaner (Michael J. Chudobiak)
* Plug leak (Stefan Kost)
* Add disable_application_handlers lockdown schema (William Jon McCann)
 
libmate 2.22.0

* Release 2.22.0
 
libmate 2.21.91

* Add schemas for tasks and calendar
* Use the GIO backend for the filechooser

libmate 2.21.90

* Set sound as enabled by default (Christian)
* Bypass a11y on win32 (Tor)
* Remove unused popt stuff (dmacks at netspace org) 
* Fix the toolbar icon size schema type (#401030)
* Win32 fixes
* SElinux fixes
* Warning fixes
* Translation updates

libmate 2.20.0

* Build fixes (Christian Persch)
* Set new background (Lucas Rocha, Vincent Untz)
* Fix module loading (Mariano Suárez-Alvarez, Christian Persch)
* Only depend on esd when esd is compiled in (Daniel Gryniewicz)
* Updated translations:
  - Albanian (Laurent Dhima)
  - Arabic (Khaled Hosny)
  - Assamese (Amitakhya Phukan)
  - Basque (Inaki Larranaga Murgoitio)
  - Bengali (Khandakar Mujahidul Islam)
  - Brazilian Portugese (Og Maciel)
  - Bulgarian (Alexander Shopov)
  - Canadian English (Adam Weinberger)
  - Catalan (Gil Forcada)
  - Danish (Kenneth Nielsen)
  - Estonian (Ivar Smolin)
  - Finnish (Ilkka Tuohela)
  - French (Stephane Raimbault)
  - Georgian (Vladimer Sichinava)
  - German (Hendrik Richter)
  - Gujarati (Ankit Patel)
  - Hungarian (Gabor Kelemen)
  - Italian (Francesco Marletta)
  - Japanese (Takeshi AIHANA)
  - Lithuanian (Žygimantas Beručka)
  - Macedonian (Jovan Naumovski)
  - Malayalam (Ani Peter)
  - Norwegian bokmål (Kjartan Maraas)
  - Polish (MATE Pl Team)
  - Portugese (Duarte Loreto)
  - Romaninan (Mişu Moldovan)
  - Russian (Nikolay V. Shmyrev)
  - Serbian (Goran Rakić)
  - Sinhala (Danishka Navin)
  - Spanish (Jorge Gonzalez)
  - Swedish (Daniel Nylander)
  - Thai (Theppitak Karoonboonyanan)
  - Ukrainian (Maxim Dziumanenko)
  - Vietnamese (Clytie Siddall)

libmate 2.19.1

* Don't initialise ALSA when sound is disabled (449249; Christian Persch)
* Translation updates: Runa Bhattacharjee (bn_IN), Hendrik Richter (de),
  Iñaki Larrañaga Murgoitio (eu), Ilkka Tuohela (fi), Ankit Patel (gu),
  Takeshi AIHANA (ja), Leonardo Fereira Fontenelle (pt_BR), Clytie Siddall (vi)

libmate 2.19.0

What's new since 2.18.0

* Rebuild the type builtins when the Makefile changes (Christian Persch)
* Make bug-buddy work with applets again (Matthias Clasen, #424949)
* Fix problems when building static libraries (Kjartan Maraas, #405352)
* Fix typos (Kenneth Nielsen, Kjartan Maraas, #429751)
* Add more information to docs for mate_program_init () (Christian Persch,
  #82943)
* Updated translations
  - mg: Thierry Randrianiriana
  - nb: Kjartan Maraas

mate-video-player/src/totem.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*- */
/* 
 * Copyright (C) 2001-2007 Bastien Nocera <hadess@hadess.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA.
 *
 * The Totem project hereby grant permission for non-gpl compatible GStreamer
 * plugins to be used and distributed together with GStreamer and Totem. This
 * permission are above and beyond the permissions granted by the GPL license
 * Totem is covered by.
 *
 * Monday 7th February 2005: Christian Schaller: Add exception clause.
 * See license_change file for details.
 *
 */

#include "config.h"

#include <glib-object.h>
#include <glib/gi18n.h>
#include <gtk/gtk.h>
#include <string.h>

#ifdef GDK_WINDOWING_X11
/* X11 headers */
#include <gdk/gdkx.h>
#include <X11/Xlib.h>
#endif

#include "totem.h"
#include "totem-private.h"
#include "totem-interface.h"
#include "totem-options.h"
#include "totem-menu.h"
#include "totem-session.h"
#include "totem-uri.h"
#include "totem-preferences.h"
#include "totem-sidebar.h"
#include "video-utils.h"

static void
long_action (void)
{
	while (gtk_events_pending ())
		gtk_main_iteration ();
}

static UniqueResponse
totem_message_received_cb (UniqueApp         *app,
			   int                command,
			   UniqueMessageData *message_data,
			   guint              time_,
			   Totem             *totem)
{
	char *url;

	if (message_data != NULL)
		url = unique_message_data_get_text (message_data);
	else
		url = NULL;

	totem_action_remote (totem, command, url);

	g_free (url);

	return UNIQUE_RESPONSE_OK;
}

static void
about_url_hook (GtkAboutDialog *about,
	        const char *link,
	        gpointer user_data)
{
	GError *error = NULL;

	if (!gtk_show_uri (gtk_widget_get_screen (GTK_WIDGET (about)),
	                   link,
	                   gtk_get_current_event_time (),
	                   &error))
	{
	        totem_interface_error (_("Could not open link"),
	                               error->message,
	                               GTK_WINDOW (about));
	        g_error_free (error);
	}
}


static void
about_email_hook (GtkAboutDialog *about,
		  const char *email_address,
		  gpointer user_data)
{
	char *escaped, *uri;

	escaped = g_uri_escape_string (email_address, NULL, FALSE);
	uri = g_strdup_printf ("mailto:%s", escaped);
	g_free (escaped);

	about_url_hook (about, uri, user_data);
	g_free (uri);
}

/* Debug log message handler: discards debug messages unless Totem is run with TOTEM_DEBUG=1.
 * If we're building in the source tree, enable debug messages by default. */
static void
debug_handler (const char *log_domain,
               GLogLevelFlags log_level,
               const char *message,
               MateConfClient *gc)
{
	static int debug = -1;

	if (debug < 0)
		debug = mateconf_client_get_bool (gc, MATECONF_PREFIX"/debug", NULL);

	if (debug)
		g_log_default_handler (log_domain, log_level, message, NULL);
}

int
main (int argc, char **argv)
{
	Totem *totem;
	MateConfClient *gc;
	GError *error = NULL;
	GOptionContext *context;
	GOptionGroup *baconoptiongroup;
	char *sidebar_pageid;

	bindtextdomain (GETTEXT_PACKAGE, MATELOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

#ifdef GDK_WINDOWING_X11
	if (XInitThreads () == 0)
	{
		gtk_init (&argc, &argv);
		g_set_application_name (_("Totem Movie Player"));
		totem_action_error_and_exit (_("Could not initialize the thread-safe libraries."), _("Verify your system installation. Totem will now exit."), NULL);
	}
#endif

	g_thread_init (NULL);
	g_type_init ();

	/* Handle command line arguments */
	context = g_option_context_new (N_("- Play movies and songs"));
	baconoptiongroup = bacon_video_widget_get_option_group();
	g_option_context_add_main_entries (context, all_options, GETTEXT_PACKAGE);
	g_option_context_set_translation_domain(context, GETTEXT_PACKAGE);
	g_option_context_add_group (context, baconoptiongroup);

	g_option_context_add_group (context, gtk_get_option_group (TRUE));
	totem_session_add_options (context);
	if (g_option_context_parse (context, &argc, &argv, &error) == FALSE) {
		g_print (_("%s\nRun '%s --help' to see a full list of available command line options.\n"),
				error->message, argv[0]);
		g_error_free (error);
	        g_option_context_free (context);
		totem_action_exit (NULL);
	}
	g_option_context_free (context);

	g_set_application_name (_("Totem Movie Player"));
	gtk_window_set_default_icon_name ("totem");
	g_setenv("PULSE_PROP_media.role", "video", TRUE);
	gtk_about_dialog_set_url_hook (about_url_hook, NULL, NULL);
	gtk_about_dialog_set_email_hook (about_email_hook, NULL, NULL);

	gc = mateconf_client_get_default ();
	if (gc == NULL)
	{
		totem_action_error_and_exit (_("Totem could not initialize the configuration engine."), _("Make sure that MATE is properly installed."), NULL);
	}

	/* Debug log handling */
	g_log_set_handler (G_LOG_DOMAIN, G_LOG_LEVEL_DEBUG, (GLogFunc) debug_handler, gc);

	/* Build the main Totem object */
	totem = g_object_new (TOTEM_TYPE_OBJECT, NULL);
	totem->gc = gc;

	/* IPC stuff */
	if (optionstate.notconnectexistingsession == FALSE) {
		totem->app = unique_app_new ("org.mate.Totem", NULL);
		totem_options_register_remote_commands (totem);
		if (unique_app_is_running (totem->app) != FALSE) {
			totem_options_process_for_server (totem->app, &optionstate);
			gdk_notify_startup_complete ();
			totem_action_exit (totem);
		} else {
			totem_options_process_early (totem, &optionstate);
		}
	} else {
		totem_options_process_early (totem, &optionstate);
	}

	/* Main window */
	totem->xml = totem_interface_load ("totem.ui", TRUE, NULL, totem);
	if (totem->xml == NULL)
		totem_action_exit (NULL);

	totem->win = GTK_WIDGET (gtk_builder_get_object (totem->xml, "totem_main_window"));

	/* Menubar */
	totem_ui_manager_setup (totem);

	/* The sidebar */
	playlist_widget_setup (totem);

	/* The rest of the widgets */
	totem->state = STATE_STOPPED;
	totem->seek = GTK_WIDGET (gtk_builder_get_object (totem->xml, "tmw_seek_hscale"));
	totem->seekadj = gtk_range_get_adjustment (GTK_RANGE (totem->seek));
	totem->volume = GTK_WIDGET (gtk_builder_get_object (totem->xml, "tmw_volume_button"));
	totem->statusbar = GTK_WIDGET (gtk_builder_get_object (totem->xml, "tmw_statusbar"));
	totem->seek_lock = FALSE;
	totem->fs = totem_fullscreen_new (GTK_WINDOW (totem->win));
	gtk_scale_button_set_adjustment (GTK_SCALE_BUTTON (totem->fs->volume),
					 gtk_scale_button_get_adjustment (GTK_SCALE_BUTTON (totem->volume)));
	gtk_range_set_adjustment (GTK_RANGE (totem->fs->seek), totem->seekadj);

	totem_session_setup (totem, argv);
	totem_setup_file_monitoring (totem);
	totem_setup_file_filters ();
	totem_setup_play_disc (totem);
	totem_callback_connect (totem);
	sidebar_pageid = totem_setup_window (totem);

	/* Show ! gtk_main_iteration trickery to show all the widgets
	 * we have so far */
	if (optionstate.fullscreen == FALSE) {
		gtk_widget_show (totem->win);
		totem_gdk_window_set_waiting_cursor (gtk_widget_get_window (totem->win));
		long_action ();
	} else {
		gtk_widget_realize (totem->win);
	}

	totem->controls_visibility = TOTEM_CONTROLS_UNDEFINED;

	/* Show ! (again) the video widget this time. */
	video_widget_create (totem);
	gtk_widget_grab_focus (GTK_WIDGET (totem->bvw));
	totem_fullscreen_set_video_widget (totem->fs, totem->bvw);

	if (optionstate.fullscreen != FALSE) {
		gtk_widget_show (totem->win);
		gdk_flush ();
		totem_action_fullscreen (totem, TRUE);
	}

	/* The prefs after the video widget is connected */
	totem_setup_preferences (totem);

	totem_setup_recent (totem);

	/* Command-line handling */
	totem_options_process_late (totem, &optionstate);

	/* Initialise all the plugins, and set the default page, in case
	 * it comes from a plugin */
	totem_object_plugins_init (totem);
	totem_sidebar_set_current_page (totem, sidebar_pageid, FALSE);
	g_free (sidebar_pageid);

	if (totem->session_restored != FALSE) {
		totem_session_restore (totem, optionstate.filenames);
	} else if (optionstate.filenames != NULL && totem_action_open_files (totem, optionstate.filenames)) {
		totem_action_play_pause (totem);
	} else {
		totem_action_set_mrl (totem, NULL, NULL);
	}

	/* Set the logo at the last minute so we won't try to show it before a video */
	bacon_video_widget_set_logo (totem->bvw, "totem");

	if (optionstate.fullscreen == FALSE)
		gdk_window_set_cursor (gtk_widget_get_window (totem->win), NULL);

	if (totem->app != NULL) {
		g_signal_connect (totem->app, "message-received",
				  G_CALLBACK (totem_message_received_cb), totem);
	}

	gtk_main ();

	return 0;
}
